README.md
# React dat.GUI

[![npm](https://img.shields.io/npm/v/react-dat-gui?style=flat-square)](https://www.npmjs.com/package/react-dat-gui)
[![Travis (.org)](https://img.shields.io/travis/claus/react-dat-gui?style=flat-square)](https://travis-ci.org/claus/react-dat-gui)
[![npm bundle size](https://img.shields.io/bundlephobia/minzip/react-dat-gui?style=flat-square)](https://bundlephobia.com/result?p=react-dat-gui)
[![GitHub](https://img.shields.io/github/license/claus/react-dat-gui?style=flat-square)](https://github.com/claus/react-dat-gui/blob/master/LICENSE)

react-dat-gui is a fully[\*](#whats-missing) featured React port of Google's esteemed [dat.GUI](https://workshop.chromeexperiments.com/examples/gui/#1--Basic-Usage) controller library. It comes packed with all of the core components you will need to cleanly integrate dat.GUIs into your React app.

<p align="middle">
  <a href="https://claus.github.io/react-dat-gui/">
    <img width="300px" src="https://i.imgur.com/1bJt59V.png" />
  </a>
  <br />
  <a href="https://claus.github.io/react-dat-gui/">Demo</a>
  &nbsp;&nbsp;&nbsp;
  <a href="https://codesandbox.io/s/react-dat-gui-emjcf?fontsize=14&module=%2Fsrc%2Fcomponents%2FReactDatGui.js">Codesandbox</a>
</p>

The dat.GUI library is designed for easily updating and interacting with objects in real time. It is used extensively in canvas and WebGL rendering demos/apps for libraries such as [three.js](http://threejs.org) and is also commonly used in browser based editing software.

## Contents

- [Basic Usage](#installation)
- [Docs](#docs)
- [Local Development](#local-development)
- [What's missing](#whats-missing)
- [Roadmap](#roadmap)

## Installation

```
npm install react-dat-gui --save
```

## Basic Usage

react-dat-gui has a wrapper component `<DatGUI />` and several control components that can be used to add functionality to the controller.

```jsx
import React from 'react';
import DatGui, { DatBoolean, DatColor, DatNumber, DatString } from 'react-dat-gui';

class App extends React.Component {
  state = {
    data: {
      package: 'react-dat-gui',
      power: 9000,
      isAwesome: true,
      feelsLike: '#2FA1D6',
    }
  }

  // Update current state with changes from controls
  handleUpdate = newData =>
    this.setState(prevState => ({
      data: { ...prevState.data, ...newData }
    }));

  render() {
    const { data } = this.state;

    return (
      <DatGui data={data} onUpdate={this.handleUpdate}>
        <DatString path='package' label='Package' />
        <DatNumber path='power' label='Power' min={9000} max={9999} step={1} />
        <DatBoolean path='isAwesome' label='Awesome?' />
        <DatColor path='feelsLike' label='Feels Like' />
      </DatGui>
    )
  }
```

## Docs

### `<DatGui />`

This is the main container component for your GUI and is the default export from the package.

#### required

| prop     | Description                                                                     | Type     |
| -------- | ------------------------------------------------------------------------------- | -------- |
| data     | The data your dat.GUI controller will mutate                                    | object   |
| onUpdate | The method which will be called whenever an update is handled by the controller | function |
| children | The dat.GUI components that make up the controller                              | array    |

#### optional

| prop       | Description                                    | Type    | Default |
| ---------- | ---------------------------------------------- | ------- | ------- |
| liveUpdate | Determines if live updates should occur        | boolean | true    |
| labelWidth | The width of the labels in any valid CSS units | string  | "40%"   |
| className  | The class name to set on the `DatGui` div      | string  | null    |
| style      | The style object to set on the `DatGui` div    | object  | null    |

### Control Components

`react-dat-gui` comes with eight built-in control components which can be used by rendering them as direct children of `<DatGui />`.

- [DatBoolean](#datboolean)
- [DatButton](#datbutton)
- [DatColor](#datcolor)
- [DatFolder](#datfolder)
- [DatPresets](#datpresets)
- [DatSelect](#datselect)
- [DatString](#datstring)

Custom control components can also be used so long as they implement the required props.

#### Common props

All child components of `<DatGui />` receive the following props implicitly, these are useful when building custom control components. See the built-in control components in [src/components](src/components) for examples of how to implement your own controls.

| prop            | Description                                                                                                                          | Type    |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------ | ------- |
| data            | The data your dat.GUI controller will mutate, the same object from `<DatGui data={data} /> | object                                  |
| labelWidth      | The width of the control name label                                                                                                  | string  |
| liveUpdate      | Determines if live updates should occur                                                                                              | boolean |
| \_onUpdateValue | A callback function for `<DatGui onUpdate={this.onUpdate} />, call this method to update dat.Gui state from your control. | function |

Below are docs for the required and optional props you can pass to each built-in control component.

##### required

- `path: string` - the path to the value within the `data` object which the component will control, eg., considering your object was `{ foo: 'bar' }`: `<DatString path='foo' />` or `{ foo: { bar: 'string' } }`: `<DatString path='foo.bar' />` for nested values.
- Note, this prop is not required for the following components
  - `DatButton`
  - `DatFolder`
  - `DatPresets`

##### optional

- `className: string` - A CSS class name
- `style: object` - A style object for inline styles
- `label: string` - The label for the controller eg., `<DatString path='message' label='Message' />`
- `labelWidth: string` - The width of the labels in any valid CSS units, overrides `<DatGUI labelWidth>`

#### `DatBoolean`

Used for controlling boolean values. Renders a checkbox input element.

#### `DatButton`

Can be used for performing any kind of function. Simply pass an `onClick` prop to the component and it will fire whenever the rendered element is clicked.

##### props

###### required

- `onClick :func` - the function to perform with the rendered element is clicked

#### `DatColor`

Uses [`react-color`](https://github.com/casesandberg/react-color/) to render a color picker component that will control color values.

#### `DatFolder`

Component which wraps other components to render them within an expandable/collapsable nested folder.

##### props

###### required

- `title: string` - The folder title eg., `<DatFolder title='MyAwesomeFolder' />`
- `children: array` - The child components to render

###### optional

- `closed: boolean` - Whether the initial state of the folder is closed, defaults to `true`

#### `DatNumber`

A number component for updating numeric values. Will render a slider if `min`, `max` and `step` props are supplied.

##### props

###### optional

- `min: number` - The minimum range for the number
- `max: number` - The maximum range for the number
- `step: number` - The amount the number should increment each tick

If your `step` prop is a float, `DatNumber` will ensure that your number field steps to the correct number of decimal places to align with the step that you've set.

#### `DatPresets`

Presets for the object which your `DatGui` is controlling can be supplied to this component as items in its `options` prop. A select field will be rendered which will allow you to easily switch between the presets.

Each item in this array will need to be in the format `{ 'presetName': ...data, ...preset }` where `...data` is your initial data and `...preset` is your preset.

##### props

###### required

- `options: array` - An array of objects, each in the format `{ 'presetName': ...data, ...preset }`

#### `DatSelect`

A select component for updating a value with one of the options supplied via the `options` prop. The initial selected value will be taken from the mapped `path` prop.

##### props

###### required

- `options: array` - A simple array of options to select from eg., `<DatSelect path='fruits' options={['apple', 'orange', 'pear']} />`

#### `DatString`

A simple text input component that can be used to mutate strings.

## Local Development

Clone the repo

```bash
git clone https://github.com/claus/react-dat-gui.git react-dat-gui
cd react-dat-gui
```

In order to see your changes to `react-dat-gui` the best way is to develop on the package and the example simultaneously. To do this, follow these steps.

### 1. Setup symlinks and install dependencies

```bash
npm install
cd example
npm install
```

### 2. Run the library in development mode

```bash
cd ..
npm run dev
```

### 3. Run the example app in development mode (in second terminal window)

```bash
cd example
npm run dev
```

After the example has compiled, it should be available for viewing at [http://localhost:3000](http://localhost:3000). Changes to the library code should now hot reload in the example app

## Scripts

| Script        | Description                                                                                                                                                 |
| ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `build`       | Builds the library for production into `/dist`                                                                                                              |
| `start`       | Starts the library in development mode with hot module reloading                                                                                            |
| `test`        | Runs unit testing suite powered by [Jest](https://github.com/facebook/jest) and [testing-library](https://github.com/testing-library/react-testing-library) |
| `lint`        | Runs linting over entire codebase with `prettier`, `eslint` and `stylelint`                                                                                 |
| `lint-js`     | Lints only javascript files                                                                                                                                 |
| `lint-styles` | Lints only stylesheet files                                                                                                                                 |
| `fix`         | Runs linting over entire codebase with `prettier`, `eslint` and `stylelint` and applies any available automatic fixes                                       |
| `fix-js`      | Lints only javascript files and applies any available automatic fixes                                                                                       |
| `fix-styles`  | Lints only stylesheet files and applies any available automatic fixes                                                                                       |
| `deploy`      | Compiles and deploys a static build of `/example` next.js app to gh-pages                                                                                   |

## What's missing

There are still a few features from the original implementation missing from this package. These are mainly related to saving and loading data as well as local storage. Animations for folder expanding/collapsing is also not currently implemented, but shouldn't be too hard to do.

For the first, I think the fact that this is now an NPM module sort of goes against it handling this sort of stuff. Google's original concept was basically a plug and play controller that could do everything if you just slam it into the browser and pass it an object. However, in module form, it's expected that you'll most likely be integrating this with an existing application. In that case, you'll probably have pretty specific needs around how you would like to save/load data into your GUI and so it's been left out for now.

Local storage however is in the roadmap and will probably be done very soon.

## Roadmap

- Loading and storing both default and preset data via `localStorage`
- Animations for `DatFolder` expanding/collapsing
- Time travel with undo/redo buttons
- ~~Better support for floating point `DatNumber`s (rounding etc.)~~ ✅

## License

[MIT](https://opensource.org/licenses/MIT)


CHANGELOG.md
# Changelog

## (2020-03-15) `v4.0.1`

Thanks to [@anselanza](https://github.com/anselanza) for their work to getting typescript types into the `dist` folder 🎉 !

### Resolves

* [#47](https://github.com/claus/react-dat-gui/issues/47) 

## (2019-09-01) `v4.0.0`

Special thanks to [@tim-soft](https://github.com/tim-soft) for this awesome [PR](https://github.com/claus/react-dat-gui/pull/33)!

### Added

- Testing framework via `jest` and `testing-library/react`
- CI integration with Travis

### Changed

- Switched deprecated `componentWillReceiveProps` to `getDerivedStateFromProps`
- All linting packages updated
- Example now uses `nextjs` and links to root modules for better dev experience
- Bundling now done with `rollup`

### Fixed

- Default preset is stored separately so that it can be reapplied after changing to other presets
- All built-in react-dat-gui components now accept className and style props
- All built-in react-dat-gui components can now have a labelWidth prop which can override the <DatGUI labelWidth> container prop and in any valid CSS units
- Automatically include styles, no longer require importing external stylesheet
- Add linting and precommit scripts that run prettier, eslint and stylelint over the codebase

## (2019-04-02) `v3.0.0`

### Changed

- Move `react`, `react-dom` and `prop-types` to peer deps [@tim-soft](https://github.com/tim-soft)

## (2019-03-01) `v2.2.0`

### Fixed

- Coding example in README [@avin](https://github.com/avin)
- `DatFolder` now passes props to children correctly once again [@rdrgn](https://github.com/rdrgn)

## (2019-01-02) `v2.0.1`

### Added

- `TypeScript` definitions
- Better support for floats

## (2019-01-02) `v2.0.0`

### Added

- `DatFolder` now accepts a `closed` boolean prop to determine if the folder should be closed or open by default

### Changed

- React v16.0.0 is now required as a peer dependency
- README updated with more docs, included a note about the React v16 peer dependency

## (2017-12-09) `v1.0.3`

### Added

- Section on React version to README
- Point regarding animations for `DatFolder` to Roadmap
- License section to README

## (2017-10-09) `v1.0.2`

### Changed

- Minor code formatting updates
- Installation guidelines

## (2017-10-09) `v1.0.1`

### Changed

- ES7 property initializers are now used to set initial state in all components rather than `componentWillMount`

## (2017-10-09) `v1.0.0`

### Added

- A lot of documentation to the README
- Deploy script to `./example` app which deploys to github pages

### Changed

- Changed default export from `Dat` to `DatGui`

## (2017-10-08) `v0.0.21`

### Changed

- Removed custom check box styling from `DatBoolean` component
- Ensured sliders are the same height as number inputs
- Minor style changes for consistency

## (2017-10-08) `v0.0.20`

### Changed

- Housekeeping in preparation for `v1.0.0`

## (2017-10-06) `v0.0.19`

### Added

- `DatPresets` component - this allows you to set presets for your DatGUI component and switch between them easily

### Changed

- Removed final reference to `reactcss`, this can be completely removed from dependencies now
- Removed unnecessary arrow functions in some `setState` callbacks
- Fixed bug where `DatNumber` input updates weren't updating sliders

## (2017-10-05) `v0.0.18`

### Added

- `DatColor` component - this allows color pickers, powered by `react-color` to be rendered for mutating colors
- Added `example:promote` script for pushing development code up from `example/src/dev` into `src`

### Changed

- Example updated with `DatColor` included

## (2017-10-04) `v0.0.17`

### Added

- `DatSelect` component

### Changed

- Example updated with `DatSelect` included

## (2017-10-03) `v0.0.16`

### Added

- Support for nested folders via `DatFolder`

### Changed

- Example `App.js` has been updated to include a nested folder example

## (2017-10-02) `v0.0.15`

### Changed

- SCSS is now broken up into partials which align with the various components

### Added

- Added `DatFolder` component
- Added this to the example

## (2017-10-02) `v0.0.14`

### Changed

- Updated component files to use `.js` file extension rather than `.jsx`
- Removed `examples` directory and created single `example` directory which is now powered by `create-react-app`
- Rewrote `webpack.config.js` so that it is compatible with Webpack v2
- Updated a bunch of dependencies in `package.json`
- Separated out `Slider` component from `DatNumber` and also added `utils` file
- Switched all event handlers in components to ES7 syntax
- Some code formatting
- Changed `build` directory to `dist`

### Added

- `example` directory
- `src/components/Slider.js`
- `src/components/utils.js`
- `src/style` directory
- `.editorconfig` file
- `.eslintrc.json` file
- `.babelrc` file
- `scripts` directory
- A number of `npm` scripts to `package.json`
- This change log!

### Removed

- The build no longer creates minified files, can add these back if required but I think it's better for the user to perform this step in their own build pipeline
- ES5 examples, don't think many people are coding with React and ES5 nowadays


dev/src/App.js
import './react-dat-gui/style/dat.css';

import DatGui, {
  DatBoolean,
  DatButton,
  DatColor,
  DatFolder,
  DatNumber,
  DatPresets,
  DatSelect,
  DatString,
} from './react-dat-gui';
import React, { Component } from 'react';

class App extends Component {
  state = {
    data: {
      string: 'Hello World',
      minMaxNumber: 66,
      number: 80,
      boolean: true,
      select: 'three',
      color: '#2FA1D6',
      random: Math.random(),
      nested: {
        string: 'Goodbye Cruel World'
      }
    },
  }

  handleUpdate = data => {
    this.setState({ data });
  }

  handleClick = () => {
    const data = {...this.state.data, random: Math.random()};

    this.setState({ data });
  }

  render() {
    const { data } = this.state;
    const swatchStyle = {
      width: '10px',
      height: '10px',
      display: 'inline-block'
    };
    const presetA = {
      string: 'Preset A',
      minMaxNumber: 33,
      number: 40,
      float: 0.001,
      boolean: false,
      select: 'one',
      color: '#e61d5f',
      random: Math.random(),
      nested: {
        string: 'Sup'
      }
    };
    const presets = [
      { 'A': { ...data, ...presetA } },
      { 'B': { ...data, string: 'Preset B' } },
      { 'C': { ...data, string: 'Preset C' } },
    ];

    return (
      <main>
        <h1>
          <b style={{color: `${data.color}`}}>react-dat-gui</b>
        </h1>
        <h2>
          Use the controls and watch your changes happen in real time!
        </h2>
        <section>
          <div>
            <b>String value:</b> {data.string}
          </div>
          <div>
            <b>Slider value:</b> {data.minMaxNumber}
          </div>
          <div>
            <b>Number value:</b> {data.number}
          </div>
          <div>
            <b>Number value is a string:</b> {(typeof data.number === 'string') ? 'true' : 'false'}
          </div>
          <div>
            <b>Float value:</b> {data.float}
          </div>
          <div>
            <b>Checkbox value:</b> {(data.boolean) ? 'true' : 'false'}
          </div>
          <div>
            <b>Select value:</b> {data.select}
          </div>
          <div>
            <b>Picked color:</b> <div style={{...swatchStyle, backgroundColor: `${data.color}`}}></div>
          </div>
          <div>
            <b>Click the button for a random number:</b> {data.random}
          </div>
          <div>
            <b>Nested string value:</b> {data.nested.string}
          </div>
        </section>
        <DatGui data={data} onUpdate={this.handleUpdate}>
          <DatPresets label='Presets' options={presets} onUpdate={this.handleUpdate} />
          <DatString path="string" label="String" />
          <DatNumber path="minMaxNumber" label="Number" min={0} max={100} step={1} />
          <DatNumber path="number" label="Number" />
          <DatNumber path="float" label="Float" min={0} max={100} step={0.001} />
          <DatNumber path="float" label="Another Float" min={0} max={100} step={0.01} />
          <DatNumber path="float" label="Min Max No Slider" min={0} max={100} step={0.01} disableSlider={true} />
          <DatBoolean path="boolean" label="Boolean" />
          <DatButton label="Button" onClick={this.handleClick} />
          <DatSelect label="Select" path='select' options={['two', 'three', 'four']}/>
          <DatColor label="Color" path='color'/>
          <DatFolder title='Folder'>
            <DatString path="string" label="String" />
            <DatNumber path="minMaxNumber" label="Number" min={0} max={100} step={1} />
            <DatFolder title='Nested Folder'>
              <DatNumber path="minMaxNumber" label="Number" min={0} max={100} step={1} />
              <DatString path="nested.string" label="String" />
              <DatFolder title="Another Nested Folder">
                <DatColor label="Color" path='color'/>
                <DatString path="nested.string" label="Nested String" />
              </DatFolder>
            </DatFolder>
          </DatFolder>
        </DatGui>
      </main>
    );
  }
}

export default App;
